indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java

```diff
@@ -3066,6 +3143,39 @@ protected abstract void updateLatestSequenceFromStream(
     return makeSequenceNumber(seq, false);
   }
 
+  /**
+   * If a task finishes reading a shard but no data was actually ingested, the task will not publish any segments.
+   * In that case, a separate message indicating that shards were closed needs to be sent to the supervisor,
+   * containing the IDs of the closed shards. The supervisor should mark those shards with the end-of-shard marker
+   * in metadata storage.
+   *
+   * @param closedShards Set of closed shards
+   * @return true if the update was successful
+   */
+  public boolean updateClosedShards(Set<PartitionIdType> closedShards)
+  {
+    // Mark partitions as closed in metadata
+    @SuppressWarnings("unchecked")
+    SeekableStreamDataSourceMetadata<PartitionIdType, SequenceOffsetType> currentMetadata =
+        (SeekableStreamDataSourceMetadata<PartitionIdType, SequenceOffsetType>) indexerMetadataStorageCoordinator.getDataSourceMetadata(
+            dataSource);
+
+    SeekableStreamDataSourceMetadata<PartitionIdType, SequenceOffsetType> cleanedMetadata =
+        createDataSourceMetadataWithClosedPartitions(currentMetadata, closedShards);
+
+    try {
+      boolean success = indexerMetadataStorageCoordinator.resetDataSourceMetadata(dataSource, cleanedMetadata);
+      if (!success) {
+        // If this fails, a subsequent task will be reassigned the closed shard and will eventually retry this.
+        log.error("Failed to update datasource metadata[%s] with expired partitions removed", cleanedMetadata);
```

Comment:
If there's a recovery mechanism, how about making this a warning instead, and including in the message a prediction of what will happen? It would help the cluster operator make sense of the log message.

Reply:
This area has been removed, the metadata commit is now handled in the main publishing logic